### PR TITLE
allow direct uploads to work within engines

### PR DESCRIPTION
**Pull request by pinzonjulian**
### Motivation / Background

This Pull Request has been created because direct uploads don't work when attempting to call them from inside a Rails Engine.

### Detail

This Pull Request changes the `convert_direct_upload_option_to_url` conditionals that determine wether it should add an upload URL as a data attribute or not. It now uses the `main_app` helper to verify if the `rails_direct_uploads_url` exists. Before this change, calling this method inside of a template in an engine would result in it being looked up in the engine's routes instead of the main app's one which it wouldn't find.

The test is there to prevent a regression.

### Additional information

No additional information

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Changes that are unrelated should be opened in separate PRs.
* [x] There are no typos in commit messages and comments.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [x] Feature branch is up-to-date with `main` (if not - rebase it).
* [x] Pull request only contains one commit for bug fixes and small features. If it's a larger feature, multiple commits are permitted but must be descriptive.
* [x] Tests are added if you fix a bug or add a feature.
* [x] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.
* [x] PR is not in a draft state.
* [x] CI is passing.

### CODE REVIEW

**Comment by pinzonjulian** [comment]
This is my first PR to Rails and I'm unsure if:
- My commit message is descriptive enough. Should I include more context like I did in the PR message?
- I should add an entry to the CHANGELOG. Even though it's a simple thing, it enables a bunch of new functionality for modular architectures using engines.

File: actionview/lib/action_view/helpers/form_tag_helper.rb [hunk 1]
```diff
@@ -1013,9 +1013,14 @@ def set_default_disable_with(value, tag_options)
         end
 
         def convert_direct_upload_option_to_url(options)
-          if options.delete(:direct_upload) && respond_to?(:rails_direct_uploads_url)
+          return options unless options.delete(:direct_upload)
+
+          if respond_to?(:main_app) && main_app.respond_to?(:rails_direct_uploads_url)
```
Comment (pinzonjulian) [inline actionview/lib/action_view/helpers/form_tag_helper.rb:1018]: In the context of just Action View, the `main_app` helper does not exist, hence these conditionals that check for its existence. The `elsif` clause is there to satisfy the tests for Action View only.

If there's a better way to do this, I'm all ears!

Comment (casperisfine) [reply actionview/lib/action_view/helpers/form_tag_helper.rb:1018]: Wouldn't it make more sense to check for `rails_direct_uploads_url` first, and then fallback to `main_app. rails_direct_uploads_url`?

First reason is that it allows you to customize that route in your engine if you wish.

Second reason is that in the vast majority of cases this code won't be executed from an engine, so might as well try the fast path first.

Comment (pinzonjulian) [reply actionview/lib/action_view/helpers/form_tag_helper.rb]: Fixed!

**Comment by casperisfine** [comment]
@pinzonjulian thanks. However two tests seem to fail consistently on CI. Can you have a look? NI'm not sure they are 100% related to your changes, but maybe you need to rebase or something like that.

**Comment by pinzonjulian** [comment]
@casperisfine I've updated the branch and everything's passed. It should be good to go

**Comment by casperisfine** [comment]
Please squash your commits.

**Comment by pinzonjulian** [comment]
Done @casperisfine !

**Pull request merged**